File: Output/java/medium/FourSumIi.java
/**https://leetcode.com/problems/4sum-ii/ */
//Given four integer arrays nums1, nums2, nums3, and nums4 all of length n, return the number of tuples (i, j, k, l) such that:
//0 <= i, j, k, l < n
//nums1[i] + nums2[j] + nums3[k] + nums4[l] == 0


class 4SumIi {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        int n = nums1.length;
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                map.put(nums1[i] + nums2[j], map.getOrDefault(nums1[i] + nums2[j], 0) + 1);
            }
        }
        int res = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                res += map.getOrDefault(-nums3[i] - nums4[j], 0);
            }
        }
        return res;
        
    }
}

    

File: Tool/PreUtil/src/output/javascript/medium/FourSumIi.js
/**https://leetcode.com/problems/4sum-ii/ */
//Given four integer arrays nums1, nums2, nums3, and nums4 all of length n, return the number of tuples (i, j, k, l) such that:
//0 <= i, j, k, l < n
//nums1[i] + nums2[j] + nums3[k] + nums4[l] == 0


/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number[]} nums3
 * @param {number[]} nums4
 * @return {number}
 */
var fourSumCount = function(nums1, nums2, nums3, nums4) {
    let n = nums1.length;
    let map = new Map();
    let result = 0;
    for(let i = 0; i < n; i++){
        for(let j = 0; j < n; j++){
            let sum = nums1[i] + nums2[j];
            map.set(sum, map.has(sum) ? map.get(sum) + 1 : 1);
        }
    }
    for(let i = 0; i < n; i++){
        for(let j = 0; j < n; j++){
            let sum = -(nums3[i] + nums4[j]);
            result += map.has(sum) ? map.get(sum) : 0;
        }
    }
    return result;
    
};

    